compiler/util/src/org/jetbrains/kotlin/utils/SmartIdentityTable.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.utils
+import com.intellij.openapi.util.Factory
+import java.util.IdentityHashMap
+
+/**
+ * [SmartIdentityTable] is a Map like structure that uses reference identity for keys.
+ * It uses 2 arrays to store keys & values until the number of entries stored is larger than 10.
+ * At that point it switches to using an IdentityHashMap.
+ *
+ * This structure can be used instead of [HashMap] when reference identity can be used and
+ * the number of entries inserted is small (<= [ARRAY_UNTIL_SIZE]) on average, drastically reducing the overhead
+ * of calls to [Object.hashCode].
+ *
+ * The implementation of [SmartIdentityTable] is not synchronized.
+ */
+class SmartIdentityTable<K, V> {
+
+    private var keysArray: MutableList<K>? = ArrayList<K>(ARRAY_UNTIL_SIZE)
+    private var valuesArray: MutableList<V>? = ArrayList<V>(ARRAY_UNTIL_SIZE)
```

Comment:
Please remove explicit `<K>` and `<V>` as per the warnings

Reply:
done

Reply:
done